README
# rcsid = $Id$

FAC is a software packge for the calculation of various atomic processes.

Contents of the package:

1. Third party content that cannot be distributed under GPL.
   These directories are bundled in fac_util.tar.gz,
   and are listed in the UtilList file.
blas/,       BLAS routines in fortran.
coul/,       fortran routines related to coulomb functions.
ionis/,      ionization balance calculation of Arnaud&Raymond,
	     and that of Mazzotta et al. Rates fitted by Verner et al.
lapack/,     fortran routines from LAPACK.
minpack/,    minpack optimization package.
quadpack/,   some routines from quadpack for quadrature.
mpfun/,      it used to be the multi-precision floating point arithmetic
             package by Beiley. Since v0.8.6, it was replaced by my own
	     implementation of quad-precision arithmetics. The associated
	     Legendre functions are also put into this directory.
ode/,	     Livermore Solver for odinary differential equations.
toms/,       some numerical routines from TOMS.

2. Everything below are distributed under GPL.
   These files are bundled in fac_core.tar.gz
README,      this file.
ChangeLog,   changes.
configure.ac,AutoConf input file
configure,   configure script after parsing configure.ac
setup.py,    python setup script for building the pfac interface
demo/,       some demo scripts.
doc/,        reference guide for FAC
doc/papers/, papers describe the theoretical backgrounds.
faclib/,     core numerical library.
python/,     python interface.
sfac/,       non-python interface.


INSTALL:
There are two interfaces available to the FAC numerical lib. One makes use
of the Python scripting language, implemented as a package "pfac".
The other interface uses a simple command parser, compiled into executables.
The two interfaces are independent. Either one or both can be installed.
One can either install from precompiled binaries or from source files.

INSTALL FROM BINARY PACKAGES:
Download appropriate binary packages. The PFAC interface is named as
FAC-version.platform.tar.gz, where version is the FAC version number
and platform is the computer platform the binary was compiled for.
Unpacking the file at some direcoty /my/own/dir will create the python package
as if python setup.py install --prefix=/my/own/dir is used. PYTHONPATH must
be set appropriately in order to use the python modeuls (see later on
installing from source).

The SFAC interface is named as SFAC-version.platform.tar.gz. This simply
contains the 3 executables under the directory bin/, sfac, scrm, and spol.
Unpack it in a directory and add bin to your PATH.

INSTALL FROM SOURCE:
STEPS:
0) download the two source packages, fac_util.tar.gz and fac_core.tar.gz,
   and gunzip, extract under the same directory.

1) ./configure
Note 1: Specify the option --with-cfortran=xxFortran, if you F77 compiler
        is not g77 or sun's f77. see faclib/cfortran.doc for supported F77
	compilers.
Note 2: If your C compiler is not gcc, then you need to supply the PIC
	(position independent code) option with --enable-cpic=***
	If your F77 compiler is not g77, then you need to supply the PIC
	option with --enable-fpic=***
	Supply the optimization option for C and F77 compilers with
	--enable-copt=*** and --enable-fopt=*** respectively if desired.
Note 3: specify --prefix=my/dir, if the default /usr/local is not what you
	want. This dir only affects the installation of SFAC executables.
Note 4: some of the functions have a parallel version using MPI. You can build
        with MPI enabled using the option --with-mpi=***, where *** is the MPI
	implementation installed on your machine. It has been tested with lammpi.
	If a different version of MPI is used, you have to supply the compile and
	link flags to the C compiler with --with-mpicompile and --with-mpilink.

2) make; make install
This installs the SFAC interface.

3) If you have Python 2.0 or later installed, then:
   make pfac
   make install-pfac
This installs the PFAC interface into Python's default site-package dir. If
this is not what you want, or you don't have the permision. then, instead of
running "make install-pfac", try:

   python setup.py install --prefix=/my/own/dir

In order for python to find the installed "pfac" modules, one would have to
define an environment variable PYTHONPATH to point to the directory
"/my/own/dir/lib/python#.#/site-packages", where "#.#" is the Python version.
In csh, this requires

    setenv PYTHONPATH /my/own/dir/lib/python#.#/site-packages

or in bash,

    PYTHONPATH=/my/own/dir/lib/python#.#/site-packages;export PYTHONPATH

Alternatively, one may do:

   python setup.py install --home=/my/own/dir

in this case, PYTHONPATH should be set to "/my/own/dir/lib/python. There
are some other ways to specify where you want the python modules go. Check
the python distutil documentation for details.

4) A binary distribution can be made by
   python setup bdist -bsfac
it will create two gzipped tar file under dist/ directory. Unpack the PFAC and
SFAC files under appropriate directories will install both sfac and pfac binary
files. Check the contents of PFAC and SFAC tar file first to determine the
directories under which they must be unpacked.

Usage:
The modules can be called just like any python modules, either from the
interpreter interactively, or pass the script to python.
e.g., python fe24_level.py.

Requirements:
The following python packages are required to run pfac.
+ numpy
+ matplotlib
+ cPickle
Also, the recent gfortran requires additional packages,
+ gfortran_linux-64 (for a linux machine)

If you are using Anaconda distribution, run
```bash
conda install numpy, matplotlib, cPickle, gfortran_linux-64
```


The non-python interface produces 3 executables, "sfac", "scrm" and
"spol", accepting input files as arguments. The syntax of the input file
for both programs is identical, and is very similar to the Python syntax,
except for the missing of flow constrols.

For more details, please read the manual, and be sure to read the FAQ
section of the manual.

Contact:
Bugs, and suggestions, please report to:
	mfgu@ssl.berkeley.edu
	Ming Feng Gu.
	Space Science Laboratory
	University of California Berkeley


ChangeLog
Version 1.1.5, Unreleased
        1. Python 3 support. 

Version 1.1.4, 11/09/2016
        1. output UTA stdev and shift for magnetic mutipole transitions as well.

Version 1.1.3, 11/10/2015
	1. fix QED corrections for UTA energies.
	2. fix UTA width and shift calculations for magnetic and even-order
	   electric mutipoles (patch by M. Krief & A. Feigel).
	3. fix configure script for modern Darwin GCC compiler (based on LLVM).

Version 1.1.2, 07/07/2015
	1. incorporate a few bug fixes from Evgeny, most notably
	   concerning the Zeemann splitting.
	2. separate the source files into fac_util and fac_core, the
	   former contains third party uitlities that cannot be
	   distributed under GPL, while the latter contains the fac core
	   that are distrubted under GPL from this version on.

Version 1.1.1, 04/30/2006
	1. change the argument of coulcc to have a small imaginary part for
	   DCOUL function. so that the results are ok for bound state energies.
	2. make low and up indexes in TransitionTable and CETable
	   interchangable.
	3. Add external magnetic and electric fields.
	4. Incident angle dependent excitation for Stark-Zeeman levels.
	5. Improve radial integrate involving continuum orbitals.
	6. Improve the construction of Y^k(r) functions.
	7. Use CPC routine for Coulomb multiple matrix elements for high
	   partial wave extrapolation.
	8. add modifiers to the configuration specification.
	9. add function to obtain expansion coefficients of G^k, useful for
	   transition rates from level to configurations.
	10.add interpolation and maxwellian rate for CI and RR.

Version 1.1.0, 10/03/2005
	1. reimplement the StructureMBPT function.
	2. add interface to njsym, and use that for general recoupling.
	3. add memory limits for the arrays used in radial.c
	4. correct an error in the 1-body contribution to autoionization.
	5. avoid overflow of 2j in UTA mode in the EN_DB file.
	6. change the format of DB_RT file.
	7. add output for the 4th NLTE workshop submission.

Version 1.0.9, 04/04/2005
	1. modify dbase.c so that it is platform independent.
	2. split the functionality of dbase.c so that it is isolated from
	   other files. It now only depends on dbase.h, and consts.h.

Version 1.0.8, 01/01/2005
	1. implement configuration average radiative transitions. including
	   energy shift and width of the UTAs.
	2. configuration average transitions for all other processes.
	3. update CRM module for UTA model.
	4. enable CI within the same non-relativisitc configuration for UTA.
	5. fix a bug in GetNumElectrons introduced ealier in this version.
	6. add JoinTable function.
	7. it seems that g77 3.4 generate incorrect code for dlamch.f with
	   -O2, now disable optimization for dlamch.f always.

Version 1.0.7, 11/10/2004
	1. add DR satellite energy to the DB_DR file.
	2. add the RRMultipole function for bound-free multipole matrix
	   elements.
	3. rewrite Angular coupling routines, much faster now.
	4. DB_TR file now stores the multipole matrix elements, instead of gf
	   values. The ASCII mode text file also has an additional column
	   giving the matrix elements.
	5. Binary output for rmatrix module.

Version 1.0.6, 09/05/2004
	1. parallelize StructureMBPT
	2. use subplx in RefineRadial.
	3. using relativistic Bethe formula for extrapolation of excitation.
	4. add an option for all plane-wave collision strength output.
	5. save the slater Y potential.
	6. improve RadialRydberg, use Dirac Coulomb for asymptotic.
	7. add an option to printout partial-wave collision strength.
	8. implement a Dirac R-matrix method.

Version 1.0.5, 05/30/2004
	1. implement magnetic sublevel ionization
	2. fix bugs in radial integration.
	3. replace PS manual with PDF format.
	4. implement a version of 2nd order MBPT for energy.
	5. fix the problem in Uheling potential.
	6. StructureMBPT, better implementation of MBPT
	7. allow the number of grid points to be changed.

Version 1.0.4, 02/10/2004
	1. make DRStrength a bit more efficient.
	2. enable the LMn inner-shell Auger extrapolation.
	3. make all atomic symbols capitalized.
	4. ReinitConfig resets the closed shells to empty.
	5. add MaxPower electron energy distribution, Maxwellian+Powerlaw.
	6. reading file for inner-shell Auger extrapolation.
	7. determine base level in structure.
	8. reimplementation of MULTI array.
	9. ensure that the wavefunction nodes are correct for high-n states.
       10. implement photoionization asymmetry parameters.

Version 1.0.3, 12/01/2003
	1. change contact information to my new address.
	2. add --enable-cpic and --enable-fpic options to configure.
	3. fix a bug in RateTable, where the optional list must be given.
	4. fix FLIBS problem for Mac OS 10.2.
	5. need to add -lcc_dynamic on Mac OS 10.3.
	6. fix the method table bug in ppol.c
	7. bug fix in IntegrateRadial function.
	8. array out-of-bound bug fix in orbital.c.

Version 1.0.2, 08/15/2003
	1. add iterative method in pfac.pol.
	2. add an optional input file in PolarizationTable to specify
	   the transitions whose polarization to be calculated.
	3. add Orientation function, which may include depolarization.
	4. fix seg. fault bug related to AvgConfig.
	5. add TRBranch and AIBranch in pfac.fac.

Version 1.0.1, 07/31/2003
	1. autoionization and capture rates between magnetic sublevels
	2. add autoionization in pfac.pol for DR satellites.
	3. change the way electron energy and density are set in pfac.pol
	4. allow calculations of DRS specifically from certain target level.

version 1.0.0, 07/15/2003
	1. add pol module for constructing magnetic sublevel population
	   models with collisional excitation.
	2. make -m and m magnetic sublevels the same in pol module.

version 0.9.9, 07/10/2003
	1. add interpolating routine for CECross.
	2. add plane-wave Born approximation for magnetic sublevel
	   excitation.
	3. change the parameter born[1] to the energy where it is
	   calculated.
	4. add the quadratic and linear interpolation feature in UVIP3P.

version 0.9.8, 06/18/2003

modqed/wk.f
c       =================================================
        subroutine wk(maxii,r,v_wk)
c       - - - - - - - - - - - - - - - - - - - - - - - - -
c       This subroutine computes Wichmann-Kroll point nuclear
c       vacuum-polarization potential by means of
C       the analytical approximation formulas from
c       A G Fainshtein, N L Manakov and A A Nekipelov,
c       J.Phys.E: At.Mol.Opt.Phys. v.23 (1990) 559-569. 
c
c       Input parameters:
c       r(i)    .... Radial grid
c       maxii   .... the size of the r array
c
c       Output parameters:
c       wk(i)   .... Wichmann-Kroll potential multiplied by r(i).
c
c       Common block parameters:
c       cl     ..... Speed of light
c       z      ..... Nuclear charge
c       ii     ..... Number of grid points
c       h, al, bt ..... Parameter of the radial grid
c       - - - - - - - - - - - - - - - - - - - - - - - - -
        implicit real*8 (a-h,o-z)
c       - - - - - - - - - - - - - - - - - - - - - - - - -
        common /z/z/cl/cl/ii/ii/h/h/al/al/bt/bt
        real*8 r(maxii),v_wk(maxii)
c       - - - - - - - - - - - - - - - - - - - - - - - - -
c       Calculation Wichmann-Kroll potential times on r
c       - - - - - - - - - - - - - - - - - - - - - - - - -
        call wk_init
c
	Do i=1,ii
          x=r(i)*cl
          v_wk(i)=u_wk(x)*r(i)
	End do
c       - - - - - - - - - - - - - - - - - - - - - - - - -
        imax=ii
        do i=ii,1,-1
	  If (dabs(v_wk(i)/r(i)).gt.1.D-40) go to 200
          imax=i
          v_wk(i)=0.d0
        enddo
c       - - - - - - - - - - - - - - - - - - - - - - - - -
 200    v_wk(ii+3)=imax
        v_wk(ii+4)=0.d0
c       - - - - - - - - - - - - - - - - - - - - - - - - -
c        open (unit=15,file='wk.dat',status='unknown')
c        write(15,'(f22.14,16x,a)') z,'z'
c        write(15,'(i22,16x,a)') ii,'ii'
c        write(15,'(d22.14,16x,a)')  h,'h'
c        write(15,'(d22.14,16x,a)') al,'al'
c        write(15,'(d22.14,16x,a)') bt,'bt'
c        write(15,*)
c        write(15,'(17x,a,23x,a,19x)') 'r','wk'
c        do i=1,ii
c          write (15,'(i6,3d24.14)') i,r(i),v_wk(i)
c        enddo
c        write(15,*)
c        write(15,'(d22.14,16x,a)') v_wk(ii+3),'wk(ii+3)'
c        write(15,'(d22.14,16x,a)') v_wk(ii+4),'wk(ii+4)'
c        close(unit=15)
c       - - - - - - - - - - - - - - - - - - - - - - - - -
        return
        end
c       =================================================
        subroutine wk_init
c       - - - - - - - - - - - - - - - - - - - - - - - - -
        implicit real*8 (a-h,o-z)
c       - - - - - - - - - - - - - - - - - - - - - - - - -
c       A G Fainshtein, N L Manakov and A A Nekipelov,
c       J.Phys.E: At.Mol.Opt.Phys. v.23 (1990) 559-569. 
c       - - - - - - - - - - - - - - - - - - - - - - - - -
c       Initialization
c       - - - - - - - - - - - - - - - - - - - - - - - - -
        real*8 f,p,q,a
        integer s
        common /wk_coeff/f(15,5),p(0:35),q(0:15),
     &  a(4,0:7,4),s(4,4)
c       - - - - - - - - - - - - - - - - - - - - - - - - -
        do m=1,15        
        do n=1,5
          f(m,n) = 0.d0
        enddo
        enddo
c       - - - - - - - - - - - - - - - - - - - - - - - - -
        p(0)=  2.094 001 968 517 539 d-02
        p(1)= -6.328 682 056 509 706 d-02
        p(2)=  1.656 675 047 469 226 d-01
        p(3)=  6.218 254 052 794 603 d-02
        p(4)=  8.561 450 880 149 488 d-01
        p(5)= -1.324 099 530 525 646 d+00
        p(6)=  1.022 894 990 055 742 d-01
        p(7)=  1.818 708 880 809 046 d-01
        p(8)=  5.002 561 255 243 687 d-04
        p(9)=  2.167 778 174 657 628 d-01
        p(10)= 5.453 835 857 080 859 d-01
        p(11)= 3.137 663 571 113 079 d-01
        p(12)=-1.619 774 880 547 481 d-01
        p(13)=-6.749 903 602 516 955 d-02
        p(14)=-2.935 144 138 820 853 d-04
        p(15)= 9.755 800 734 714 044 d-02
        p(16)=-1.994 016 822 313 688 d-01
        p(17)=-4.777 623 414 220 549 d-02
        p(18)= 7.503 472 520 308 676 d-03
        p(19)= 3.440 742 164 594 281 d-05

        p(20)= 2.829 420 9807 d-03
        p(21)= 1.419 551 8237 d-04
        p(22)= 1.348 983 3978 d-05
        p(23)= 2.013 400 8556 d-07
        p(24)= 8.602 918 4536 d-06
        p(25)=-1.787 016 5724 d-05
        p(26)= 2.326 960 7856 d-05
        p(27)=-1.779 846 3760 d-05
        p(28)= 8.727 474 4996 d-06
        p(29)=-2.923 560 6110 d-06
        p(30)= 6.892 936 0802 d-07
        p(31)=-1.149 169 9126 d-07
        p(32)= 1.330 283 5807 d-08
        p(33)=-1.019 098 8753 d-09
        p(34)= 4.650 804 1566 d-11
        p(35)=-9.578 592 4366 d-13
c       - - - - - - - - - - - - - - - - - - - - - - - - -
        q(0)  = 2.09400223235d-02
        q(1)  = 1.42425723499d-02
        q(2)  = 0.76419630941d-02
        q(3)  = 0.36124746165d-02
        q(4)  = 0.15118743043d-02
        q(5)  = 0.05247168246d-02
        q(6)  = 0.01797758276d-02
        q(7)  =-0.04761316284d-02
        q(8)  = 0.19713924639d-02
        q(9)  =-0.62421312791d-02
        q(10) = 1.29259217859d-02
        q(11) =-1.89507086023d-02
        q(12) = 1.89089237003d-02
        q(13) =-1.24611625668d-02
        q(14) = 0.48793792307d-02
        q(15) =-0.08858438257d-02
c       - - - - - - - - - - - - - - - - - - - - - - - - -
        f(5,1) = 0.3812769d+00
        f(5,2) = 0.0266795d+00
        f(5,3) = 0.0000860d+00

        f(6,1) = 0.4946161d+00
        f(6,2) = 0.0458093d+00
        f(6,3) = 0.0002918d+00
        f(6,4) = 6.d-08

        f(7,1) = 0.6150374d+00
        f(7,2) = 0.0707397d+00
        f(7,3) = 0.0007216d+00
        f(7,4) = 5.d-07

        f(8,1) = 0.7412290d+00
        f(8,2) = 0.1014082d+00
        f(8,3) = 0.0014705d+00
        f(8,4) = 23.0d-07
        f(8,5) = 1.0d-10

        f(9,1) =0.8721898d+00
        f(9,2) =0.1376368d+00
        f(9,3) =0.0026288d+00
        f(9,4) =69.d-07
        f(9,5) =2.0d-09

        f(10,1) =1.0071432d+00
        f(10,2) =0.1791838d+00
        f(10,3) =0.0042776d+00
        f(10,4) =167.0d-07
        f(10,5) =8.0d-09

        f(11,1) = 1.1454776d+00
        f(11,2) = 0.2557770d+00
        f(11,3) = 0.0064866d+00
        f(11,4) = 346.0d-07
        f(11,5) = 3.0d-08


        f(12,1) = 1.2867042d+00
        f(12,2) = 0.2771342d+00
        f(12,3) = 0.0093134d+00
        f(12,4) = 643.0d-07
        f(12,5) = 9.0d-08

        f(13,1) = 1.4304276d+00
        f(13,1) = 0.3329753d+00
        f(13,1) = 0.0128043d+00
        f(13,1) = 1097.0d-07
        f(13,1) = 2.0d-07

        f(14,1) = 1.5763244d+00
        f(14,1) = 0.3930296d+00
        f(14,1) = 0.0169953d+00
        f(14,1) = 1750.0d-07
        f(14,1) = 5.0d-07

        f(15,1) = 1.7241274d+00
        f(15,1) = 0.4570401d+00
        f(15,1) = 0.0219132d+00
        f(15,1) = 2644.0d-07
        f(15,1) = 9.0d-07
c       - - - - - - - - - - - - - - - - - - - - - - - - -      
        do i=1,4
        do m=0,7
        do n=1,4
          a(i,m,n)=0.d0
        enddo
        enddo
        enddo
c       - - - - - - - - - - - - - - - - - - - - - - - - -     
        a(1,0,1) = 2.9675d+00
        a(1,0,2) = 7.8894d-01
        a(1,0,3) =-3.0201d-04
        a(1,0,4) = 4.9725d-04

        a(1,1,1) =-4.4288d+00
        a(1,1,2) =-2.9946d+01
        a(1,1,3) =-9.0441d-01
        a(1,1,4) = 2.4826d+00

        a(1,2,1) = 5.9339d01
        a(1,2,2) = 8.4881d+02
        a(1,2,3) = 5.8388d+01
        a(1,2,4) =-1.0745d+02

        a(1,3,1) =-5.6820d+02
        a(1,3,2) =-1.4888d+04
        a(1,3,3) =-1.5899d+03
        a(1,3,4) = 2.7128d+03

        a(1,4,1) = 4.6131d+03
        a(1,4,2) = 1.6003d+05
        a(1,4,3) = 2.2880d+04
        a(1,4,4) =-3.8249d+04

        a(1,5,1) =-2.6282d+04
        a(1,5,2) =-1.0259d+06
        a(1,5,3) =-1.8036d+05
        a(1,5,4) = 3.0100d+05

        a(1,6,1) = 8.9448d+04
        a(1,6,2) = 3.5980d+06
        a(1,6,3) = 7.3716d+05
        a(1,6,4) =-1.2368d+06

        a(1,7,1) =-1.3412d+05
        a(1,7,2) =-5.3067d+06
        a(1,7,3) =-1.2228d+06
        a(1,7,4) = 2.0678d+06
c       - - - - - - - - - - - - - - - - - - - - - - - - -     
        a(2,0,1) =-6.64673d-01
        a(2,0,2) = 4.68647d-05
        a(2,0,3) = 2.56839d-02

        a(2,1,1) = 5.70540d-01
        a(2,1,2) =-9.83836d-04
        a(2,1,3) =-3.21021d-03

        a(2,2,1) =-9.43955d-01
        a(2,2,2) = 8.28746d-03
        a(2,2,3) = 1.01163d-02

        a(2,3,1) = 1.18405d+00
        a(2,3,2) = 5.94372d-02
        a(2,3,3) =-1.62093d-02

        a(2,4,1) =-8.91689d-01
        a(2,4,2) =-5.31431d-02
        a(2,4,3) = 1.17762d-02

        a(2,5,1) = 3.64707d-01
        a(2,5,2) = 7.52653d-03
        a(2,5,3) =-4.69814d-03

        a(2,6,1) =-6.17902d-02
        a(2,6,2) =-4.54081d-04
        a(2,6,3) = 7.76469d-04
c       - - - - - - - - - - - - - - - - - - - - - - - - -
        a(3,0,1) = 1.93743d-02
        a(3,0,2) = 1.64381d-01
        a(3,0,3) = 2.53250d-02

        a(3,1,1) =-6.43399d-01
        a(3,1,2) =-3.05149d-01
        a(3,1,3) = 3.56651d-03

        a(3,2,1) = 2.56565d-01
        a(3,2,2) = 2.31560d-01
        a(3,2,3) =-8.14980d-03

        a(3,3,1) =-1.09235d-01
        a(3,3,2) =-9.49948d-02
        a(3,3,3) = 6.21168d-03

        a(3,4,1) = 4.14226d-02
        a(3,4,2) = 2.22450d-02
        a(3,4,3) =-3.29527d-03

        a(3,5,1) =-7.28744d-03
        a(3,5,2) =-2.80168d-03
        a(3,5,3) = 7.60885d-04

        a(3,6,1) = 4.54914d-04
        a(3,6,2) = 1.47382d-04
        a(3,6,3) =-6.03993d-05
c       - - - - - - - - - - - - - - - - - - - - - - - - -     
        a(4,0,1) =-2.4734758d+09
        a(4,0,2) =-1.5115262d+04

        a(4,1,1) = 3.2270784d+09
        a(4,1,2) = 1.4945586d+04

        a(4,2,1) =-1.7207915d+09
        a(4,2,2) =-5.8361389d+03

        a(4,3,1) = 4.7635740d+08
        a(4,3,2) = 1.1476935d+03

        a(4,4,1) =-7.1414759d+07
        a(4,4,2) =-1.2098290d+02

        a(4,5,1) = 5.4305859d+06
        a(4,5,2) = 6.5439613d+00

        a(4,6,1) =-1.6479928d+05
        a(4,6,2) =-1.4296421d-01
c       - - - - - - - - - - - - - - - - - - - - - - - - -
        s(1,1) = 1
        s(1,2) = 1
        s(1,3) = 0
        s(1,4) = 0

        s(2,1) = 0
        s(2,2) =-3
        s(2,3) = 0
        s(2,4) = 0

        s(3,1) =-1
        s(3,2) = 2
        s(3,3) = 0
        s(3,4) = 0

        s(4,1) =-13
        s(4,2) =-6
        s(4,3) = 0
        s(4,4) = 0
c       - - - - - - - - - - - - - - - - - - - - - - - - -
        return
        end
c       =================================================
        real*8 function u_wk(x)
c       - - - - - - - - - - - - - - - - - - - - - - - - -
        implicit real*8 (a-h,o-z)
c       - - - - - - - - - - - - - - - - - - - - - - - - -
c       A G Fainshtein, N L Manakov and A A Nekipelov,
c       J.Phys.E: At.Mol.Opt.Phys. v.23 (1990) 559-569. 
c       - - - - - - - - - - - - - - - - - - - - - - - - -
        common /z/z/cl/cl
        real*8 break (5)
        parameter (pi=3.1415926535 8979323846d0)
        real*8 f,p,q,a
        integer s
        common /wk_coeff/f(15,5),p(0:35),q(0:15),
     &  a(4,0:7,4),s(4,4)
c       - - - - - - - - - - - - - - - - - - - - - - - - -
        alz=z/cl
        dlam=1.d0-dsqrt(1.d0-alz)*dsqrt(1.d0+alz)
c       - - - - - - - - - - - - - - - - - - - - - - - - -      
        if (x.lt.10.d0) then
          i=1
          if (x.gt.4.d0) then
            i=4
          else
            if (x.gt.1.5d0) then
              i=3
            else
              if (x.gt.0.15d0) i=2
            endif
          endif
c
          v3=0.d0
          if (x.le.4.25d0) then   
            do n=0,8
              v3=v3+p(n)*x**(n-1)
            enddo
            do n=0,5
              v3=v3+p(n+9)*x**(n+2)*dlog(x)
            enddo
            do n=0,4
              v3=v3+p(n+15)*x**(n+3)*dlog(x)**2
            enddo
          else
            do n=0,15
              v3=v3+p(n+20)*(1.d1/x)**(2*n)
            enddo
            v3=v3/x**5
          endif
          v3=v3/cl*alz**3
c
          u_wk=v3
c
          if (x.lt.0.15d0) then
            q5=0.d0
            do n=1,15                              
              q5=q5+q(n)*dlam**n
            enddo
            q5=q5*alz**3/cl
            u_wk=u_wk+q5/x
          endif
c
          sum=1.d0
          do n=1,4
          do m=0,7
            sum=sum+a(i,m,n)*(dlam**n)*x**(m+s(i,n))
          enddo
          enddo
          u_wk=u_wk/sum
c
        else
c
          u_wk=2.d0/225.d0/x**5+59.d0/1323.d0/x**7+(1977.d0+
     &    20.d0*alz*alz)/4725.d0/x**9+(1258680.d0+34144.d0*alz*alz)/
     &    190575/x**11+(1960420032.d0+93618070.d0*alz*alz+
     &    96096.d0*alz**4)/12297285.d0/x**13
          u_wk=u_wk*alz**3    
c      
          prod=576.d0/x**13
          do m=5,15
            prod=prod*m*m/(x*x)
            sum=0.d0
            do n=1,5
              sum=sum+f(m,n)*alz**(2*n+1)
            enddo
            u_wk=u_wk+sum*prod
          enddo
          u_wk=u_wk/(pi*cl)
        endif
c       - - - - - - - - - - - - - - - - - - - - - - - - -      
        return
        end
